rdtools/system_analysis.py

```diff
@@ -0,0 +1,587 @@
+        Calculate clear-sky ambient temperature and store in self.clearsky_ambient_temperature
+        '''
+        times = self.clearsky_poa.index
+        if self.pvlib_location is None:
+            raise ValueError('pvlib location must be provided')
+        loc = self.pvlib_location
+
+        cs_amb_temp = clearsky_temperature.get_clearsky_tamb(times, loc.latitude, loc.longitude)
+
+        self.clearsky_ambient_temperature = cs_amb_temp
+
+    def pvwatts_norm(self, poa, cell_temperature):
+        '''
+        Normalize PV energy to that expected from a PVWatts model.
+
+        Parameters
+        ---------
+        poa: plane of array irradiance in W/m^2 (numeric)
+        cell_temperature: cell temperature in Celsius (numeric)
+
+        Returns
+        -------
+        (Normalized pv energy, associated insolation)
+        '''
+        if self.pv_nameplate is None:
+            renorm = True
+            pv_nameplate = 1.0
+        else:
+            renorm = False
+
+        if self.temperature_coefficient is None:
+            raise ValueError('Temperature coeffcient must be available to perform pvwatts_norm')
+
+        pvwatts_kws = {"poa_global": poa,
+                       "P_ref": pv_nameplate,
+                       "T_cell": cell_temperature,
+                       "G_ref": 1000,
+                       "T_ref": 25,
+                       "gamma_pdc": self.temperature_coefficient}
+
+        normalized, insolation = normalization.normalize_with_pvwatts(self.pv_energy, pvwatts_kws)
+
+        if renorm:
+            # Normalize to the 95th percentile for convenience, this is renormalized out
+            # in the calculations but is relevant to normalized_filter()
+            x = normalized[np.isfinite(normalized)]
+            normalized = normalized / x.quantile(0.95)
+
+        return normalized, insolation
+
+    def filter(self, normalized, case):
+        '''
+        Calculate filters based on those in rdtools.filtering. Uses self.filter_params, which is a dict,
+        the keys of which are names of functions in rdtools.filtering, and the values of which are dicts
+        containing the associated parameters with which to run the filtering functions. See examples for
+        details on how to modify filter parameters.
+
+        Parameters
+        ----------
+        normalized: normalized PV energy (Pandas Time Series)
+        case: 'sensor' or 'clearsky' which filtering protocol to apply. Affects whether filtering.csi_filter()
+               is used and whether result is stored in self.sensor_filter or self.clearsky_filter)
+
+        Returns
+        -------
+        None
+        '''
+        bool_filter = True
+
+        if case == 'sensor':
+            poa = self.poa
+            cell_temp = self.cell_temperature
+        if case == 'clearsky':
+            poa = self.clearsky_poa
+            cell_temp = self.clearsky_cell_temperature
+
+        if 'normalized_filter' in self.filter_params.keys():
```

Comment:
Here and below, `key in my_dict` is semantically the same as `key in my_dict.keys()`, although the former avoids generating a temporary `dict_keys` object.